File: Assignment-7/ocaml/statistics.ml
let read_ints_from_file filename =
  let channel = open_in filename in
  let rec loop acc =
    try
      let value = Scanf.fscanf channel " %d" (fun x -> x) in
      loop (value :: acc)
    with End_of_file ->
      close_in channel;
      List.rev acc
  in
  loop []

let mean numbers =
  let total = List.fold_left ( + ) 0 numbers in
  float_of_int total /. float_of_int (List.length numbers)

let median numbers =
  let sorted = List.sort compare numbers in
  let n = List.length sorted in
  if n mod 2 = 1 then
    float_of_int (List.nth sorted (n / 2))
  else
    let left = List.nth sorted ((n / 2) - 1) in
    let right = List.nth sorted (n / 2) in
    (float_of_int left +. float_of_int right) /. 2.0

let mode numbers =
  let sorted = List.sort compare numbers in
  let rec count_runs lst current_value current_count acc =
    match lst with
    | [] -> List.rev ((current_value, current_count) :: acc)
    | x :: xs ->
        if x = current_value then
          count_runs xs current_value (current_count + 1) acc
        else
          count_runs xs x 1 ((current_value, current_count) :: acc)
  in
  match sorted with
  | [] -> ([], 0)
  | x :: xs ->
      let frequencies = count_runs xs x 1 [] in
      let max_count = List.fold_left (fun m (_, c) -> max m c) 0 frequencies in
      if max_count = 1 then
        ([], 1)
      else
        let modes =
          frequencies
          |> List.filter (fun (_, c) -> c = max_count)
          |> List.map fst
        in
        (modes, max_count)

let join_ints values =
  values
  |> List.map string_of_int
  |> String.concat ", "

let () =
  let values = read_ints_from_file "../tests/sample_input.txt" in
  let count = List.length values in
  if count = 0 then (
    print_endline "No integers found in input.";
    exit 1
  );

  Printf.printf "OCaml Functional Statistics\n";
  Printf.printf "Input count: %d\n" count;
  Printf.printf "Mean: %.2f\n" (mean values);
  Printf.printf "Median: %.2f\n" (median values);

  let modes, frequency = mode values in
  match modes with
  | [] ->
      if frequency = 1 then
        print_endline "Mode: no mode (all values unique)"
      else
        print_endline "Mode: no data"
  | _ ->
      Printf.printf "Mode: %s (frequency %d)\n" (join_ints modes) frequency

File: Assignment-7/c/statistics.c
#include <stdio.h>
#include <stdlib.h>

static int compare_ints(const void *a, const void *b) {
    int x = *(const int *)a;
    int y = *(const int *)b;
    return (x > y) - (x < y);
}

double calculate_mean(const int *arr, int n) {
    long long sum = 0;
    int i;
    for (i = 0; i < n; i++) {
        sum += arr[i];
    }
    return (double)sum / (double)n;
}

double calculate_median(int *arr, int n) {
    qsort(arr, n, sizeof(int), compare_ints);

    if (n % 2 == 1) {
        return (double)arr[n / 2];
    }
    return ((double)arr[(n / 2) - 1] + (double)arr[n / 2]) / 2.0;
}

void calculate_mode(const int *arr, int n) {
    int *copy = (int *)malloc((size_t)n * sizeof(int));
    int i, max_freq = 0, current_freq = 1;
    int first = 1;

    if (copy == NULL) {
        printf("Mode: memory allocation failed\n");
        return;
    }

    for (i = 0; i < n; i++) {
        copy[i] = arr[i];
    }
    qsort(copy, n, sizeof(int), compare_ints);

    for (i = 1; i <= n; i++) {
        if (i < n && copy[i] == copy[i - 1]) {
            current_freq++;
        } else {
            if (current_freq > max_freq) {
                max_freq = current_freq;
            }
            current_freq = 1;
        }
    }

    if (max_freq == 1) {
        printf("Mode: no mode (all values unique)\n");
        free(copy);
        return;
    }

    printf("Mode: ");
    current_freq = 1;
    for (i = 1; i <= n; i++) {
        if (i < n && copy[i] == copy[i - 1]) {
            current_freq++;
        } else {
            if (current_freq == max_freq) {
                if (!first) {
                    printf(", ");
                }
                printf("%d", copy[i - 1]);
                first = 0;
            }
            current_freq = 1;
        }
    }
    printf(" (frequency %d)\n", max_freq);

    free(copy);
}

int main(void) {
    FILE *fp = fopen("../tests/sample_input.txt", "r");
    int capacity = 16;
    int count = 0;
    int *values;
    int *median_copy;
    int i;

    if (fp == NULL) {
        printf("Error: could not open ../tests/sample_input.txt\n");
        return 1;
    }

    values = (int *)malloc((size_t)capacity * sizeof(int));
    if (values == NULL) {
        fclose(fp);
        printf("Error: memory allocation failed\n");
        return 1;
    }

    while (1) {
        int x;
        if (fscanf(fp, "%d", &x) != 1) {
            break;
        }
        if (count == capacity) {
            int *tmp;
            capacity *= 2;
            tmp = (int *)realloc(values, (size_t)capacity * sizeof(int));
            if (tmp == NULL) {
                free(values);
                fclose(fp);
                printf("Error: memory reallocation failed\n");
                return 1;
            }
            values = tmp;
        }
        values[count++] = x;
    }
    fclose(fp);

    if (count == 0) {
        free(values);
        printf("No integers found in input.\n");
        return 1;
    }

    median_copy = (int *)malloc((size_t)count * sizeof(int));
    if (median_copy == NULL) {
        free(values);
        printf("Error: memory allocation failed\n");
        return 1;
    }

    for (i = 0; i < count; i++) {
        median_copy[i] = values[i];
    }

    printf("C Procedural Statistics\n");
    printf("Input count: %d\n", count);
    printf("Mean: %.2f\n", calculate_mean(values, count));
    printf("Median: %.2f\n", calculate_median(median_copy, count));
    calculate_mode(values, count);

    free(median_copy);
    free(values);
    return 0;
}
